### PR TITLE
WIP: gerrit: Use `base` to upload exact revisions

### LINKED ISSUE #9364: gerrit: Utilize `base` option in `gerrit upload`
I recently contributed [support for more gerrit push options](https://github.com/jj-vcs/jj/pull/9353).

There is now only one (non google-internal) push option not yet supported by jj: [base](https://gerrit-review.googlesource.com/Documentation/user-upload.html#base). I want to share some design ideas for this option.

The `base` option can be used to tell gerrit for which pushed commits it should create changes. It defaults to all branch heads, meaning gerrit will create changes for all pushed commits not yet in any branch.

One common use case for me is when stacking changes on top of coworker's changes. I might rebase locally and when pushing only want to push my changes, without pushing the coworker's changes I rebased on my machine. This can be accomplished my setting `base` to the commit hash of the change I'm stacking on.

There are multiple possibilities how `base` could be supported by jj, with different trade-offs.

We could either implement a `--base` option, that accepts a revset, converts it to a commit hash and passes that as `base` to gerrit. This would be relatively easy to implement and mirror the git workflow. Also manually figuring out which `base` you need to pass to achieve the intended result is not always that easy for new users.

Another (in my opinion better) option would be to integrate this with the existing `--revision` option:

```
    /// The revset, selecting which revisions are sent in to Gerrit
    ///
    /// This can be any arbitrary set of commits. Note that when you push a
    /// commit at the head of a stack, all ancestors are pushed too. This means
    /// that `jj gerrit upload -r foo` is equivalent to `jj gerrit upload -r
    /// 'mutable()::foo`.
    ///
    /// If this is not provided, it will check whether @ has a description.
    /// * If it does, it will upload @
    /// * Otherwise, it will upload @-
```

If we are willing to break backwards compatibility we could define `--revision` to no longer push the mentioned commits and their ancestors, but only the mentioned commits by automatically setting `base` to the parents of the mentioned commits. Then `gerrit upload -r foo` would no longer upload `mutable()::foo`, but only upload `foo` by pushing `foo` and setting `base=foo-` (I would however keep the default behaviour when no `-r` is given as `-r mutable()::@`). This would also support calls  such as `gerrit upload -r x::y`, which would otherwise (if x is a descendant of mutable()) also be identical to `gerrit upload -r mutable()::y` .

As far as I know jj enforces a 6 month deprecation policy, which would make this a bit hard, as this would not be removing a feature, but changing the behaviour of an existing feature. I'm not quite sure what the best approach for this would be.

CC: @thoughtpolice @matts1 (because you seem to have written the majority of the gerrit code)

### PR DESCRIPTION
**Pull request by freundTech**
Closes #9364.

Work in progress.

This adds the changes proposed in #9364. Because they are breaking changes they are hidden behind a config option for now. The default value of this config option could then be flipped after a deprecation period.

This still needs a bit more testing and probably a few edge case fixes. Opening as draft PR to enable design discussions. Not yet for code review (I know there's a bunch of stuff to fix).

# Checklist

If applicable:

- [ ] I have updated `CHANGELOG.md`
- [ ] I have updated the documentation (`README.md`, `docs/`, `demos/`)
- [ ] I have updated the config schema (`cli/src/config-schema.json`)
- [ ] I have added/updated tests to cover my changes
- [ ] I fully understand the code that I am submitting (what it does,
      how it works, how it's organized), including any code drafted by an LLM.
- [ ] For any prose generated by an LLM, I have proof-read and copy-edited with
      an eye towards deleting anything that is irrelevant, clarifying anything
      that is confusing, and adding details that are relevant. This includes,
      for example, commit descriptions, PR descriptions, and code comments.

### CODE REVIEW

**Comment by PhilipMetzger** [comment]
please expand on the commit motivations and move the fixes line into the one which "completes" the work.

File: cli/src/config/misc.toml
Comment (PhilipMetzger) [inline cli/src/config/misc.toml:1]: nit: this is missing a `config.md` entry explaining what it does.

File: cli/src/commands/gerrit/upload.rs [hunk 1]
```diff
@@ -469,19 +508,25 @@ pub async fn cmd_gerrit_upload(
         return Ok(());
     }
 
-    // If you have the changes main -> A -> B, and then run `jj gerrit upload -r B`,
-    // then that uploads both A and B. Thus, we need to ensure that A also
-    // has a Change-ID.
-    // We make an assumption here that all immutable commits already have a
-    // Change-ID.
```
Comment (PhilipMetzger) [inline cli/src/commands/gerrit/upload.rs:476]: nit: its probably useful to preserve that comment

File: cli/src/commands/gerrit/upload.rs [hunk 2]
```diff
@@ -405,6 +408,37 @@ fn push_options(args: &UploadArgs) -> Result<Vec<String>, CommandError> {
...
+) -> IndexResult<Vec<CommitId>> {
+    let index = repo.index();
+    let root_commit = repo.store().root_commit_id();
+
+    let mut bases = Vec::new();
+    let mut visited = HashSet::new();
+    let mut queue = VecDeque::from([commit]);
+
+    while let Some(current_commit) = queue.pop_front() {
+        let parents = index.parents(&current_commit)?;
```
Comment (PhilipMetzger) [inline cli/src/commands/gerrit/upload.rs:424]: nit: this results in N index lookups/calls which can be expensive, so offering a stream-based implementation would definitely better for Google and ERSC. cc @martinvonz for an opinion.

File: lib/src/index.rs [hunk 3]
```diff
@@ -140,6 +140,9 @@ pub trait Index: Send + Sync {
     /// `candidates` list it will appear at most once in the output.
     fn heads(&self, candidates: &mut dyn Iterator<Item = &CommitId>) -> IndexResult<Vec<CommitId>>;
 
+    /// Returns the parents of commit ID `commit_id`
+    fn parents(&self, commit_id: &CommitId) -> IndexResult<Vec<CommitId>>;
```
Comment (PhilipMetzger) [inline lib/src/index.rs:144]: nit: This is related to my other comment, but maybe a variant which takes an iterator of CommitIds is better here for alternate backends which aren't Git and then also returns a stream.

**Review by PhilipMetzger — COMMENTED** [review]
minor pass.
